cajita/src/Cajita_Splines.hpp

```diff
@@ -593,172 +600,203 @@ struct SplineData<Scalar, Order, EntityType, SplineDataMemberTypes<Tags...>>
-KOKKOS_INLINE_FUNCTION std::enable_if_t<
-    SplineData<Scalar, Order, EntityType, DataTags>::has_physical_distance>
-setSplineData( SplinePhysicalDistance,
-               SplineData<Scalar, Order, EntityType, DataTags>& data,
-               const Scalar low_x[3], const Scalar p[3], const Scalar dx[3] )
+template <typename Scalar, int Order, std::size_t NumSpaceDim, class EntityType,
+          class DataTags>
+KOKKOS_INLINE_FUNCTION
+    std::enable_if_t<SplineData<Scalar, Order, NumSpaceDim, EntityType,
+                                DataTags>::has_physical_distance>
+    setSplineData(
+        SplinePhysicalDistance,
+        SplineData<Scalar, Order, NumSpaceDim, EntityType, DataTags>& data,
+        const Scalar low_x[NumSpaceDim], const Scalar p[NumSpaceDim],
+        const Scalar dx[NumSpaceDim] )
 
 {
-    using spline_type =
-        typename SplineData<Scalar, Order, EntityType, DataTags>::spline_type;
+    using spline_type = typename SplineData<Scalar, Order, NumSpaceDim,
+                                            EntityType, DataTags>::spline_type;
     Scalar offset;
-    for ( int d = 0; d < 3; ++d )
+    for ( std::size_t d = 0; d < NumSpaceDim; ++d )
     {
         offset = low_x[d] - p[d];
         for ( int n = 0; n < spline_type::num_knot; ++n )
             data.d[d][n] = offset + data.s[d][n] * dx[d];
     }
 }
 
-template <typename Scalar, int Order, class EntityType, class DataTags>
-KOKKOS_INLINE_FUNCTION std::enable_if_t<
-    !SplineData<Scalar, Order, EntityType, DataTags>::has_physical_distance>
+template <typename Scalar, int Order, std::size_t NumSpaceDim, class EntityType,
+          class DataTags>
+KOKKOS_INLINE_FUNCTION std::enable_if_t<!SplineData<
+    Scalar, Order, NumSpaceDim, EntityType, DataTags>::has_physical_distance>
 setSplineData( SplinePhysicalDistance,
-               SplineData<Scalar, Order, EntityType, DataTags>&,
-               const Scalar[3], const Scalar[3], const Scalar[3] )
+               SplineData<Scalar, Order, NumSpaceDim, EntityType, DataTags>&,
+               const Scalar[NumSpaceDim], const Scalar[NumSpaceDim],
+               const Scalar[NumSpaceDim] )
 {
 }
 
 //---------------------------------------------------------------------------//
 // Evaluate spline data at a point in a uniform mesh.
-template <typename Scalar, int Order, class Device, class EntityType,
-          class DataTags>
-KOKKOS_INLINE_FUNCTION void
-evaluateSpline( const LocalMesh<Device, UniformMesh<Scalar>>& local_mesh,
-                const Scalar p[3],
-                SplineData<Scalar, Order, EntityType, DataTags>& data )
+template <typename Scalar, int Order, std::size_t NumSpaceDim, class Device,
+          class EntityType, class DataTags>
+KOKKOS_INLINE_FUNCTION void evaluateSpline(
+    const LocalMesh<Device, UniformMesh<Scalar, NumSpaceDim>>& local_mesh,
+    const Scalar p[NumSpaceDim],
+    SplineData<Scalar, Order, NumSpaceDim, EntityType, DataTags>& data )
```

Comment:
I think the main issue here is the `NumSpaceDim` template in `SplineData` in terms of backwards compatibility - the rest of the objects here are implementation details. I don't think we can move this to the back of the list because `DataTags` is variadic and thus will have a variable length. I would argue the best way to use `SplineData` is to always have it be a generic template in your function. I think we need a `createSplineData` function that would return `SplineData` with the proper types via deduction and a user can just assign this to `auto`. 

Ultimately this doesn't fix the backwards compatibility issue so I think we might need to break it on this one.

@streeve comments?

Reply:
Well we already missed one class for full backwards compatibility, so I think one more is fine. Unless you want to rename and wrap the existing with an explicit 3.

Reply:
we have discussed this and are going to move forward with the current structure.